src/useHumanbound.ts

```diff
@@ -0,0 +1,116 @@
+import { BigNumber, providers } from "ethers";
+import { useCallback, useEffect, useState } from "react";
+
+import { AlchemyNetworkNames, Web3ChainReference, humanboundDeployments } from "./deployments";
+import { IHumanboundToken, IHumanboundToken__factory } from "./types";
+
+type PartialRecord<K extends keyof any, T> = {
+  [P in K]?: T;
+};
+type AlchemyAPIKeyConfig = PartialRecord<
+  | Web3ChainReference.EIP155_ETHEREUM_MAINNET
+  | Web3ChainReference.EIP155_ETHEREUM_RINKEBY
+  | Web3ChainReference.EIP155_ETHEREUM_KOVAN
+  | Web3ChainReference.EIP155_OPTIMISM
+  | Web3ChainReference.EIP155_OPTIMISM_GOERLI
+  | Web3ChainReference.EIP155_ARBITRUM_ONE
+  | Web3ChainReference.EIP155_ARBITRUM_GOERLI
+  | Web3ChainReference.EIP155_POLYGON_MAINNET
+  | Web3ChainReference.EIP155_POLYGON_MUMBAI,
+  string
+>;
+
+/**
+ * useHumanbound react hook for humanbound token data
+ * Provides access to three functions:
+ *  - hasHBT: @param `address`
+ *  - getHBTIdOfOwner: @param `address`
+ *  - getOwnerOf: @param `tokenId`
+ *
+ * Usage:
+ * const { hasHBT, getHBTIdOfOwner, getOwnerOf } = useHumanbound(alchemyConfig);
+ *
+ * `alchemyConfig` is a configuration for your Alchemy node API keys.
+ * You must provide a map of networks to API keys for the supported networks above.
+ * These networks are all optional.
+ */
+export const useHumanbound = (alchemyConfig: AlchemyAPIKeyConfig) => {
+  const [humanboundContract, setHumanboundContract] = useState<IHumanboundToken>();
+
+  const ethereum = (window as any).ethereum;
+  if (!ethereum) throw new Error("useHumanbound: your browser does not support injected web3 provider");
+
+  // Initialise chainId as the initial provider chain
+  const [chainId, setChainId] = useState<number>(ethereum.networkVersion);
+
+  // Setup provider listener to update chainId if a chain switch occurs
+  useEffect(() => {
+    ethereum?.on("chainChanged", (chainId: string) => {
+      setChainId(parseInt(chainId, 16));
+    });
+  }, []);
+
+  // Update the contract instance if the chain switches
+  useEffect(() => {
+    if (ethereum.networkVersion === 0) return;
+    if (!Object.values(Web3ChainReference).find(value => value == chainId)) {
+      console.error(`useHumanbound: unsuppported network ${chainId}`);
+      return;
+    }
+    if (!alchemyConfig[chainId as Web3ChainReference])
+      throw new Error(`useHumanbound: missing apikey for chainId ${chainId}`);
+
+    const provider = new providers.JsonRpcProvider(
+      `https://${AlchemyNetworkNames[chainId as Web3ChainReference]}.g.alchemy.com/v2/${
+        alchemyConfig[chainId as Web3ChainReference]
```

Comment:
Yeh... I don't think it's much of an issue from a user perspective since it's abstracted away but in any case is there a better way to do this whilst keeping the type?